### PR TITLE
Allow 0.5 seconds difference in file timestamps before warning of conflicting change

### PR DESCRIPTION
**Pull request by takluyver**
See gh-484

### CODE REVIEW

File: notebook/static/notebook/js/notebook.js [hunk 1]
```diff
@@ -2721,7 +2721,10 @@ define([
             return this.contents.get(this.notebook_path, {content: false}).then(
                 function (data) {
                     var last_modified = new Date(data.last_modified);
-                    if (last_modified > that.last_modified) {
+                    // We want to check last_modified (disk) > that.last_modified (our last save)
+                    // In some cases the filesystem reports an inconsistent time,
+                    // so we allow 0.5 seconds difference before complaining.
+                    if ((last_modified.getTime() - that.last_modified.getTime()) > 500) {  // 500 ms
```
Comment (blink1073) [inline notebook/static/notebook/js/notebook.js]: Use `Math.abs()` to allow it to go in either direction?

Comment (blink1073) [inline notebook/static/notebook/js/notebook.js]: Actually, reading the original comment, I take that back.

Comment (takluyver) [inline notebook/static/notebook/js/notebook.js]: I don't think we need it to go in either direction - we want to show the message if the on-disk timestamp is newer than the timestamp from when we last saved it.  The other way round would be weird.

**Comment by lcnature** [comment]
Hi, I tried to install in editable mode. It seems that even after I have left the tab for a while, if I press Ctrl+s to save the notebook, it still throws the warning of whether I want to overwrite change outside this tab. 
Is there any log I can pull out to help understand this issue:

This is my conda environment:
bleach                    2.0.0                     <pip>
cycler                    0.10.0                   py36_0  
Cython                    0.26                      <pip>
dbus                      1.10.20                       0  
decorator                 4.1.2                     <pip>
entrypoints               0.2.3                     <pip>
expat                     2.1.0                         0  
fontconfig                2.12.1                        3  
freetype                  2.5.5                         2  
glib                      2.50.2                        1  
gst-plugins-base          1.8.0                         0  
gstreamer                 1.8.0                         0  
html5lib                  0.999999999               <pip>
icu                       54.1                          0  
ipykernel                 4.6.1                     <pip>
ipython                   6.1.0                     <pip>
ipython-genutils          0.2.0                     <pip>
jedi                      0.10.2                    <pip>
Jinja2                    2.9.6                     <pip>
jpeg                      9b                            0  
jsonschema                2.6.0                     <pip>
jupyter-client            5.1.0                     <pip>
jupyter-core              4.3.0                     <pip>
libffi                    3.2.1                         1  
libgcc                    5.2.0                         0  
libgfortran               3.0.0                         1  
libiconv                  1.14                          0  
libpng                    1.6.27                        0  
libxcb                    1.12                          1  
libxml2                   2.9.4                         0  
MarkupSafe                1.0                       <pip>
matplotlib                2.0.2               np113py36_0  
mistune                   0.7.4                     <pip>
mkl                       2017.0.3                      0  
mpi4py                    2.0.0                     <pip>
nbconvert                 5.2.1                     <pip>
nbformat                  4.3.0                     <pip>
nibabel                   2.1.0                     <pip>
nilearn                   0.3.1                     <pip>
nitime                    0.7                       <pip>
nodejs                    6.11.0                        0    conda-forge
numpy                     1.13.1                   py36_0  
openssl                   1.0.2l                        0  
pandocfilters             1.4.1                     <pip>
pcre                      8.39                          1  
pexpect                   4.2.1                     <pip>
pickleshare               0.7.4                     <pip>
pip                       9.0.1                    py36_1  
prompt-toolkit            1.0.15                    <pip>
ptyprocess                0.5.2                     <pip>
pybind11                  2.1.1                     <pip>
Pygments                  2.2.0                     <pip>
pymanopt                  0.2.2                     <pip>
pyparsing                 2.2.0                     <pip>
pyparsing                 2.1.4                    py36_0  
pyqt                      5.6.0                    py36_2  
python                    3.6.2                         0  
python-dateutil           2.6.1                    py36_0  
python-dateutil           2.6.1                     <pip>
pytz                      2017.2                   py36_0  
pyzmq                     16.0.2                    <pip>
qt                        5.6.2                         5  
readline                  6.2                           2  
scikit-learn              0.18.2              np113py36_0  
scipy                     0.19.1              np113py36_0  
setuptools                27.2.0                   py36_0  
simplegeneric             0.8.1                     <pip>
sip                       4.18                     py36_0  
six                       1.10.0                    <pip>
six                       1.10.0                   py36_0  
sqlite                    3.13.0                        0  
terminado                 0.6                       <pip>
testpath                  0.3.1                     <pip>
Theano                    0.9.0                     <pip>
tk                        8.5.18                        0  
tornado                   4.5.1                     <pip>
traitlets                 4.3.2                     <pip>
typing                    3.6.1                     <pip>
wcwidth                   0.1.7                     <pip>
webencodings              0.5.1                     <pip>
wheel                     0.29.0                   py36_0  
xz                        5.2.2                         1  
zlib                      1.2.8                         3

**Comment by takluyver** [comment]
Can you try adding this in to the Javascript just before the check, and see in the browser's console how different the two timestamps are:

```javascript
console.log(last_modified);
console.log(that.last_modified);
```

**Comment by lcnature** [comment]
Sorry, it was my mistake. I installed master branch instead of from your repo. So I just incorporated the change to my master branch copy and no waning is thrown now in the tab. 
I turned on debug level logging and here is what I got in the terminal when I pressed Ctr+s to save notebook. Is there anything wrong?

```[D 11:10:31.764 NotebookApp] 200 GET /api/contents/plosCB_2.ipynb?content=0&_=1501168198769 (127.0.0.1) 3.01ms
[I 11:10:31.779 NotebookApp] Saving file at /plosCB_2.ipynb
[D 11:10:31.780 NotebookApp] Saving /mnt/scratch/RSA_Bayesian/scripts/plosCB_2.ipynb
[D 11:10:31.817 NotebookApp] copystat on /mnt/scratch/RSA_Bayesian/scripts/.~plosCB_2.ipynb failed
    Traceback (most recent call last):
      File "/home/mingbo/jupyter-source/notebook/notebook/services/contents/fileio.py", line 53, in copy2_safe
        shutil.copystat(src, dst)
      File "/home/mingbo/anaconda3/envs/test_jupyter/lib/python3.6/shutil.py", line 225, in copystat
        _copyxattr(src, dst, follow_symlinks=follow)
      File "/home/mingbo/anaconda3/envs/test_jupyter/lib/python3.6/shutil.py", line 157, in _copyxattr
        names = os.listxattr(src, follow_symlinks=follow_symlinks)
    OSError: [Errno 22] Invalid argument: '/mnt/scratch/RSA_Bayesian/scripts/plosCB_2.ipynb'
[D 11:10:31.887 NotebookApp] 200 PUT /api/contents/plosCB_2.ipynb (127.0.0.1) 109.43ms
[D 11:10:31.916 NotebookApp] copystat on /mnt/scratch/RSA_Bayesian/scripts/.ipynb_checkpoints/plosCB_2-checkpoint.ipynb failed
    Traceback (most recent call last):
      File "/home/mingbo/jupyter-source/notebook/notebook/services/contents/fileio.py", line 53, in copy2_safe
        shutil.copystat(src, dst)
      File "/home/mingbo/anaconda3/envs/test_jupyter/lib/python3.6/shutil.py", line 225, in copystat
        _copyxattr(src, dst, follow_symlinks=follow)
      File "/home/mingbo/anaconda3/envs/test_jupyter/lib/python3.6/shutil.py", line 157, in _copyxattr
        names = os.listxattr(src, follow_symlinks=follow_symlinks)
    OSError: [Errno 22] Invalid argument: '/mnt/scratch/RSA_Bayesian/scripts/plosCB_2.ipynb'
[D 11:10:31.917 NotebookApp] 201 POST /api/contents/plosCB_2.ipynb/checkpoints (127.0.0.1) 19.21ms
```

**Comment by takluyver** [comment]
Thanks, that sounds like it's working correctly.

The errors you see are something separate, though probably also a consequence of the shared filesystem. They're already being caught and logged, so it should be OK.

**Comment by gnestor** [comment]
Excellent @takluyver!!

**Comment by pluviosilla** [comment]
I still get this error occasionally when opening a notebook I've downloaded from Github or Kaggle. Here's a quick fix if you need relief and don't have time to go on a safari hunt for a solution: `%autosave 0`

It doesn't fix the problem and you still will get a dozen error messages every time you save the notebook, but at least you can work without error dialogs popping up every 4 or 5 seconds.

**Comment by brettbrocato** [comment]
Something closely-related is still a challenge in 5.7.0:  https://github.com/ipython/ipython/issues/8456

**Pull request merged**